### ISSUE TITLE
Widget

### ISSUE THREAD

**Issue by samuelclay** [samuelclay/turntouch-ios#12]
Now that we have action buttons (#4), let's make a widget! 

https://developer.apple.com/design/human-interface-guidelines/ios/extensions/widgets/

If possible, have it mirror the app (4 modes, 4 buttons) and allow switching between the four apps (3 apps if we're in 12 button mode).

**Comment by samuelclay** [comment]
If there's no interaction allowed, then display all four apps with 4 diamonds one above the other?

**Comment by Dejal** [comment]
Should be fun (and useful). Interaction should be allowed, e.g. PCalc's widget supports tapping buttons to perform calculations. May need a bit of refactoring to support accessing and performing actions from an extension, though.

**Comment by Dejal** [comment]
I have added a new [`widget`](https://github.com/samuelclay/turntouch-ios/tree/widget) branch with the current work-in-progress on the widget.

It isn't finished yet, but the widget is available, and the **Phone** and **Music** modes are functional. Other modes will follow in the future.  For now, unsupported modes are replaced with the **Phone** one.

If you want to try it, note that you'll need to set your developer team in the **Widget Extension** target in the project.  The **Dejal** app and widget targets are ones I can build, since I don't have access to your personal developer account.

Here's a screenshot of the collapsed appearance (a bit tight):

![IMG_9916](https://user-images.githubusercontent.com/1318849/80779752-70bf7a80-8b21-11ea-9283-ded693f1d486.png)

And the expanded appearance:

![IMG_9917](https://user-images.githubusercontent.com/1318849/80779777-82a11d80-8b21-11ea-9b2c-7951771842b3.png)

**Comment by samuelclay** [comment]
Screenshots looking great and show lots of promise. 

I'm having trouble getting the widget to show up in notification center. I changed the app group to the one I registered online, changed the team, and compiled just fine, but no widget. It's being embedded (and not signed) but it's not showing up for some reason. Any ideas?

**Comment by Dejal** [comment]
You shouldn't have had to change the app group, though may have needed to click the reload button below it.  Though maybe you would have needed to, if linked to the developer account; I'm not sure.

If you changed the app group name from `group.com.turntouch.ios-remote`, some code will need to be correspondingly changed.

Here are what my signing settings look like for the `Dejal` targets, if that helps; the app:

<img width="714" alt="Screen Shot 2020-05-01 at 08 23 25" src="https://user-images.githubusercontent.com/1318849/80816865-2bcd2f80-8b85-11ea-9311-66a8ba1920b9.png">

The widget:

<img width="735" alt="Screen Shot 2020-05-01 at 08 23 15" src="https://user-images.githubusercontent.com/1318849/80816876-3091e380-8b85-11ea-82fc-38dc739a4074.png">

Note that the widget won't show up by default; you have to scroll to the bottom of the widgets and click **Edit**, and add the widget.

**Comment by samuelclay** [comment]
Yeah I changed both of those app groups to the one I already had registered (`group.com.turntouch.appgroup`). And I've been checking the bottom of the Edit screen.

**Comment by Dejal** [comment]
Fair enough. Make sure you change the code to use that group, too.  And commit your changes once you get it working, so I don't stomp on them when I resume working on it (no urgency; won't be until the last week of May).

That won't affect the widget appearing, though.

Ah, I notice that the `iOS Deployment Target` in the widget's `Build Settings` is set to 13.4; if you're using an older iOS version, that could explain it.  That is too high. The app is iOS 12, so you can change it to match that.

If that doesn't solve it, try deleting the app from the phone and re-installing.

**Comment by samuelclay** [comment]
Yep, I was on 13.3.X and not 13.4. Upgraded and it wasn't there, but then I launched the app and it was immediately there. So the app needs to launch first before it's discoverable. 

Looks great! We should probably tint the background 30% on tap, since the tiny blue outline isn't doing much (and let's remove that blue color as well).

Can we lower the font size more in miniature mode?

**Comment by Dejal** [comment]
I'm glad that solved it.

Yeah, need some visual tweaks. I'll note those.

**Comment by Dejal** [comment]
More progress on the widget, including the UI tweaks, and support for opening the app to handle actions that can't be done in the widget (due to requiring state or UI not supported there).

A few more modes remaining to do before completion.

**Comment by samuelclay** [comment]
Excellent, I love that this is the season of widgets.

**Comment by Dejal** [comment]
WWDC20 has deprecated the current widget mechanism, in favor of a [SwiftUI-based non-interactive design](https://developer.apple.com/widgets/).

For Turn Touch, this is problematic. We wouldn't be able to switch modes in the widget — or if we did, we'd have to have the widget open the app to do so. Similarly, tapping any action would have to open the app to perform it. Several of the modes already do that, since they have state that can't be done in the current widget, but it'd have to do that even for the simplest actions like changing the volume.

The widget would also have to be rewritten using SwiftUI.  The current implementation is still supported for now, but is limited to the bottom of the Today view instead of the home screen on iOS 14, and will probably stop working before long.

So I'm not sure it's worthwhile to continue working on this?

**Comment by samuelclay** [comment]
Yeah let's hold off on this for now. If I see other IoT widgets doing it then maybe we'll find a way.

**Comment by Dejal** [comment]
I was very disappointed by this change. I can see why Apple wanted to go with the timeline-based model (like for Apple Watch complications), so they don't have to wait for things to load, but surely iPhones are powerful enough to run a few widgets in the background if necessary, or other solutions that allow real functionality. Several other kinds of widget won't be possible anymore, e.g. calculators (e.g. PCalc), podcast players (e.g. Overcast), etc. Hopefully they'll reconsider this in a future iOS version.

I won't do any more work on this for now.
